File: wrappers/javascript/src/_circuit_breaker.ts
/**
 * Minimal circuit breaker for control-plane calls. Three states:
 *
 *   - **closed**: requests flow normally; consecutive failures count.
 *   - **open**: every request fails fast without hitting the network
 *     until `resetAfterMs` elapses.
 *   - **half-open**: one probe request is allowed; success closes the
 *     circuit, failure re-opens it.
 *
 * Scoped per-instance so the batcher and the public-key registrar can
 * share one instance (or use separate ones), and so tests can construct
 * a breaker with a tight timeout.
 */

/** State transitions observable from diagnostics / logs. */
export type CircuitState = "closed" | "open" | "half_open";

/** Options for {@link CircuitBreaker}. */
export interface CircuitBreakerOptions {
  /** Consecutive failures before the circuit opens. Default: 5. */
  failureThreshold?: number;
  /** Base time the circuit stays open before admitting a probe. Default: 30_000 ms. */
  resetAfterMs?: number;
  /**
   * Maximum random jitter added on top of {@link resetAfterMs} when the
   * circuit opens. Each open-event picks a fresh value in
   * ``[0, resetJitterMs]`` so independent breakers tripped by a shared
   * outage do not all probe at the same instant — a thundering herd
   * that would re-overwhelm a recovering control plane and trip every
   * breaker again immediately. Default: 5_000 ms (~17 % of the base).
   *
   * Pattern: AWS Builders Library "exponential backoff and jitter",
   * applied to the reset window instead of the retry delay.
   * Set to ``0`` to opt out (not recommended in production).
   */
  resetJitterMs?: number;
  /** Override of the clock; test-only. */
  now?: () => number;
  /**
   * Override of the random source. Test-only — production callers
   * use ``Math.random``. Pass a function returning a value in
   * ``[0, 1)`` to produce deterministic jitter for assertions.
   */
  random?: () => number;
}

/** Diagnostic snapshot. */
export interface CircuitBreakerDiagnostics {
  state: CircuitState;
  consecutiveFailures: number;
  openedAt: number | null;
  /**
   * Time (ms) the breaker will remain open from {@link openedAt}
   * before admitting a half-open probe. Equals ``resetAfterMs`` plus
   * the jitter rolled at the most recent open-event. ``null`` when
   * the breaker has never opened in this process.
   */
  effectiveResetMs: number | null;
}

/**
 * Small, synchronous circuit breaker. Instances are not thread-safe
 * in the abstract, but JS's single-threaded event loop makes them
 * trivially safe per-isolate. Callers wire `allow()` before each
 * attempt and `recordSuccess()` / `recordFailure()` after.
 */
export class CircuitBreaker {
  private readonly failureThreshold: number;
  private readonly resetAfterMs: number;
  private readonly resetJitterMs: number;
  private readonly now: () => number;
  private readonly random: () => number;

  private state: CircuitState = "closed";
  private consecutiveFailures = 0;
  private openedAt: number | null = null;
  // Effective reset window for the current open-event = resetAfterMs +
  // a fresh random sample in [0, resetJitterMs]. Re-rolled every time
  // the circuit transitions into ``open`` so two consecutive opens of
  // the same breaker have different recovery windows. Null while the
  // circuit is closed (no jitter has been rolled yet).
  private currentResetMs: number | null = null;

  constructor(opts: CircuitBreakerOptions = {}) {
    this.failureThreshold = opts.failureThreshold ?? 5;
    this.resetAfterMs = opts.resetAfterMs ?? 30_000;
    this.resetJitterMs = opts.resetJitterMs ?? 5_000;
    if (this.resetJitterMs < 0) {
      throw new Error(
        `CircuitBreaker.resetJitterMs must be >= 0; got ${this.resetJitterMs.toString()}`,
      );
    }
    this.now = opts.now ?? ((): number => Date.now());
    this.random = opts.random ?? ((): number => Math.random());
  }

  /**
   * Called before each outbound attempt. Returns `true` if the attempt
   * should proceed, `false` if the circuit is open. Transitions from
   * `open` to `half_open` when the (jittered) reset window has elapsed.
   */
  allow(): boolean {
    if (this.state === "closed") return true;
    if (this.state === "half_open") return true;
    const now = this.now();
    const reset = this.currentResetMs ?? this.resetAfterMs;
    if (this.openedAt !== null && now - this.openedAt >= reset) {
      this.state = "half_open";
      return true;
    }
    return false;
  }

  /** Record a successful attempt. Resets the failure counter. */
  recordSuccess(): void {
    this.consecutiveFailures = 0;
    this.state = "closed";
    this.openedAt = null;
    this.currentResetMs = null;
  }

  /** Record a failed attempt. May open the circuit. */
  recordFailure(): void {
    this.consecutiveFailures += 1;
    if (
      this.state === "half_open" ||
      this.consecutiveFailures >= this.failureThreshold
    ) {
      this.state = "open";
      this.openedAt = this.now();
      // Roll fresh jitter on every open-event so two breakers tripped
      // by the same outage do not probe at the same instant. Math.random
      // returns in [0, 1) so the addition lies in [0, resetJitterMs).
      // Short-circuit when jitter is 0 — the random source is never
      // called, which keeps the explicit-opt-out path observable in
      // tests that mock ``random`` to throw.
      const jitter =
        this.resetJitterMs === 0
          ? 0
          : Math.floor(this.random() * this.resetJitterMs);
      this.currentResetMs = this.resetAfterMs + jitter;
    }
  }

  /** Current state for diagnostics. */
  diagnostics(): CircuitBreakerDiagnostics {
    return {
      state: this.state,
      consecutiveFailures: this.consecutiveFailures,
      openedAt: this.openedAt,
      effectiveResetMs: this.currentResetMs,
    };
  }

  /** Test-only reset. */
  reset(): void {
    this.state = "closed";
    this.consecutiveFailures = 0;
    this.openedAt = null;
    this.currentResetMs = null;
  }
}

File: wrappers/python/src/checkrd/_circuit_breaker.py
"""Circuit breaker for control-plane HTTP calls.

Mirrors `wrappers/javascript/src/_circuit_breaker.ts` byte-for-byte —
same three states (``closed`` / ``open`` / ``half_open``), same
defaults (5 consecutive failures opens; 30 s reset window). One
breaker instance can be shared across the batcher and the public-key
registrar so a single sustained outage of the control plane fast-fails
both call sites instead of every caller racing through its own retry
budget.

Why a circuit breaker matters:

  - Without one, a control-plane outage means every caller pays the
    full retry budget on every attempt. With ``DEFAULT_MAX_ATTEMPTS=3``
    and ``DEFAULT_MAX_SLEEP_SECS=8``, that's ~24 seconds of latency on
    every batcher flush during an outage — unacceptable in async hot
    paths.
  - Open → half-open → closed transitions let the breaker probe the
    control plane periodically; one successful probe closes the
    circuit. No background thread or polling required.
  - Per-instance state means tests can construct a breaker with tight
    thresholds and a mock clock.

Thread safety:
  Uses :class:`threading.Lock` because the batcher worker thread and
  the foreground caller can both call ``allow()`` / ``record_*`` from
  different threads. The lock is uncontended in steady state — only
  the failure-burst windows take it, so the cost is negligible.
"""

from __future__ import annotations

import threading
import time
from dataclasses import dataclass
from typing import Callable, Literal, Optional


#: Three states observable from diagnostics / logs. Closed is the
#: happy path; open means fast-fail; half_open means "let one probe
#: through to see if the control plane is back".
CircuitState = Literal["closed", "open", "half_open"]


@dataclass(frozen=True)
class CircuitBreakerDiagnostics:
    """Snapshot of breaker state for ``Checkrd.healthy()`` exposure.

    Mirrors the JS SDK's ``CircuitBreakerDiagnostics`` interface so a
    multi-runtime fleet exports identically-shaped fields.
    """

    state: CircuitState
    consecutive_failures: int
    opened_at: Optional[float]


class CircuitBreaker:
    """Three-state circuit breaker with thread-safe accounting.

    Wire ``allow()`` before each outbound attempt and
    ``record_success()`` / ``record_failure()`` after::

        if not breaker.allow():
            raise APIConnectionError(message="circuit open")
        try:
            response = make_request(...)
            breaker.record_success()
        except SomeFailure:
            breaker.record_failure()
            raise

    Defaults match the JS SDK so a fleet running both runtimes has
    identical breaker behavior; override per call site for tests.

    Args:
        failure_threshold: Consecutive failures before the circuit
            opens. Default 5 — tuned to ride out a momentary 5xx burst
            without cracking, but to fail fast on a sustained outage.
        reset_after_secs: Time the circuit stays open before admitting
            a probe. Default 30 s — short enough that a brief outage
            recovers quickly, long enough that we don't hammer a still-
            unhealthy backend.
        now: Optional clock override for tests.
    """

    def __init__(
        self,
        *,
        failure_threshold: int = 5,
        reset_after_secs: float = 30.0,
        now: Optional[Callable[[], float]] = None,
    ) -> None:
        self._failure_threshold = failure_threshold
        self._reset_after_secs = reset_after_secs
        self._now = now if now is not None else time.monotonic
        self._lock = threading.Lock()
        self._state: CircuitState = "closed"
        self._consecutive_failures = 0
        self._opened_at: Optional[float] = None

    def allow(self) -> bool:
        """Return True if the next attempt should proceed.

        Transitions ``open → half_open`` automatically when the reset
        window has elapsed. The half-open probe is admitted; the
        outcome of that probe (recorded via :meth:`record_success` or
        :meth:`record_failure`) decides whether the circuit closes
        again or re-opens.
        """
        with self._lock:
            if self._state == "closed":
                return True
            if self._state == "half_open":
                return True
            # state == "open"
            if (
                self._opened_at is not None
                and self._now() - self._opened_at >= self._reset_after_secs
            ):
                self._state = "half_open"
                return True
            return False

    def record_success(self) -> None:
        """Mark the most recent attempt as successful.

        Closes the circuit and resets the failure counter. Safe to
        call even when already closed — the only state change is the
        counter reset, which is idempotent.
        """
        with self._lock:
            self._consecutive_failures = 0
            self._state = "closed"
            self._opened_at = None

    def record_failure(self) -> None:
        """Mark the most recent attempt as failed.

        Opens the circuit if either:
        - The half-open probe failed (one strike — back to open), or
        - The consecutive-failure count crosses the threshold.

        The "half-open probe failed" branch is the safety hatch for a
        control plane that flaps repeatedly: each probe failure
        re-arms the reset timer, so we don't probe in a tight loop.
        """
        with self._lock:
            self._consecutive_failures += 1
            if (
                self._state == "half_open"
                or self._consecutive_failures >= self._failure_threshold
            ):
                self._state = "open"
                self._opened_at = self._now()

    def diagnostics(self) -> CircuitBreakerDiagnostics:
        """Snapshot of breaker state for monitoring and ``healthy()``."""
        with self._lock:
            return CircuitBreakerDiagnostics(
                state=self._state,
                consecutive_failures=self._consecutive_failures,
                opened_at=self._opened_at,
            )

    def reset(self) -> None:
        """Force-close the circuit. Test-only — production callers should
        not need to invoke this; the half-open probe handles recovery."""
        with self._lock:
            self._state = "closed"
            self._consecutive_failures = 0
            self._opened_at = None


__all__ = [
    "CircuitBreaker",
    "CircuitBreakerDiagnostics",
    "CircuitState",
]
